service.rb

```diff
@@ -79,4 +76,15 @@ def daily_data_form
       </form>
     STR
   end
+
+  def store(daily_data)
+    store = YAML::Store.new("daily_data.yml")
+    store.transaction do
+      store[:all_data] << daily_data
+    end
+  end
+
+  def daily_data
+    YAML.load_file("daily_data.yml")[:all_data]
```

Comment:
This works fine, but that’s kind of by luck rather than by design. Ideally we wouldn’t be relying on any of the details of how `YAML::Store` chooses to represent an array of hashes on disk: `YAML.load_file("daily_data.yml")[:all_data]` does the right thing today but maybe it will break in future if the implementation details of `YAML::Store` change.

I realise that sounds like a purely hypothetical concern — there’s really only a single obvious way of serialising an array of hashes as YAML, so why would it ever change? — but the larger principle is that we should [avoid relying on implementation details](https://en.wikipedia.org/wiki/Dependency_inversion_principle) when possible. If we can treat `YAML::Store` as a completely opaque abstraction for both storing data and reading it back again later, we’ll be insulated from any under-the-hood changes.

Another way of looking at it is: if we only rely on the API provided by `YAML::Store` and don’t care about how it works, it will be easier to replace it with a different class later if we want to. For example, [`PStore`](https://rubydoc.info/stdlib/pstore/PStore) is API-compatible with `YAML::Store` but [uses `Marshal.dump` and `Marshal.load`](https://github.com/ruby/pstore/blob/v0.1.1/lib/pstore.rb#L475-L485) to persist binary representations of Ruby objects instead of YAML, which is faster and more space-efficient at the cost of human-readability and portability between different platforms and languages, so we might want to switch to that at some point if performance or disk space become important. This kind of flexibility, enabled by intentional ignorance of the underlying details, is an important property of maintainable software.

So: can you rewrite `daily_data` to use `YAML::Store`?

Reply:
It seems like `PStore` is different from `IO` since there is no `IO#transaction`. Why not use `IO` type objects instead, since it's a more general detail? Is it because `PStore` stores hashes instead of just string? 

How will I know for the future whether or not which details are API compatible, or is it something you kind of learn over time?

Reply:
Can you say a little more about what you mean? I don’t fully understand the question.